04-store-api/starter/controllers/products.js

```diff
@@ -0,0 +1,83 @@
+const Product = require("../models/product");
+//////////////////////////////////////////////////////////////////STATIC/////////////////////////////////////////////////////////////////////////////
+
+const getAllProductsStatic = async (req, res) => {
+  const search = "a";
+  const products = await Product.find({ price: { $gt: 100 } })
+    .sort("name")
+    .select("name price")
+    .limit(10)
+    .skip(1).numericFilters;
+  // throw new Error("Throw async errors");
+  res.status(200).json({ products, nbHits: products.length });
+};
+
+//////////////////////////////////////////////////////////////////
+const getAllProducts = async (req, res) => {
+  console.log(req.query);
+  const { featured, company, name, sort, fields, numericFilters } = req.query;
+  const queryObject = {};
+  if (featured) {
+    queryObject.featured = featured === "true" ? "true" : "false";
+  }
+  if (company) {
+    queryObject.company = company;
+  }
+  if (name) {
+    queryObject.name = { $regex: name, $options: "i" };
+  }
+  if (numericFilters) {
+    const operatorMap = {
+      ">": "$gt",
+      ">=": "$gte",
+      "<": "$lt",
+      "<=": "$lte",
+      "=": "$eq",
+    };
+    const regEx = /\b(<|>|>=|=|<|<=)\b/g;
+    let filters = numericFilters.replace(
+      regEx,
+      (match) => `-${operatorMap[match]}-`
+    );
+    const options = ["price", "rating"];
+    filters = filters.split(",").forEach((element) => {
+      const [field, operator, value] = element.split("-");
+      if (options.includes(field)) {
+        queryObject[field] = { [operator]: Number(value) }
+      }
+    });
```

Comment:
Requires change:
This has a bug where it overwrites properties, for example this query: 
```?numericFilters=price>10,price<20,rating=4.5``` 
will output a filter: 
```{ price: { $lt: 20 }, rating: { $eq: 4.5 } }``` 
instead of
```{ price: { $gt: 10, $lt: 20 }, rating: { $eq: 4.5 } }```

The fix is to spread out any previous properties in the filter and then use parseFloat --- I'd recommend using "parseFloat" here since `rating` is a decimal value (4.5) and Mongo can handle floats for `price` even though it's an integer (10, 20):
```js
queryObject[field] = { ...queryObject[field], [operator]: parseFloat(value) } }
```

That said, all of this logic can be simplified to...
```js
if (numericFilters) {
    const operatorMap = {
        '>': '$gt',
        '>=': '$gte',
        '=': '$eq',
        '<': '$lt',
        '<=': '$lte',
    }

    const allowedFilterProps = {
        'price': 'price',
        'rating': 'rating'
    }

    // first, split the query into filters: [ "price>10", "price<20", "rating=4.5" ], 
    // then iterate over each filter
    numericFilters.split(",").forEach(filter => {
        // sanitize filter by keeping Aa-Zz, 0-9, <, =, >, . 
        const santizedFilter = filter.replace(/[^\w<=>\.]/g, '')

        // find the matching operator
        const operator = santizedFilter.match(/\b(<|>|>=|=|<=)\b/g)

        // match operator to Mongo string
        const allowedOperator = operatorMap[operator]

        // split filter via operator: "price>20" becomes ["price", "20] 
        const [prop, value] = santizedFilter.split(operator)

        // match prop to allowed filter property
        const allowedFilter = allowedFilterProps[prop]

        if (allowedFilter && allowedOperator && value) {
            // assign filter to query: { price: { $gt: 10, $lt: 20 }, rating: { $eq: 4.5 } }
            queryObject[allowedFilter] = { ...queryObject[allowedFilter], [allowedOperator]: parseFloat(value) }
        }
    })
}
```